lib/graphql/schema/traversal.rb

```diff
@@ -87,17 +103,30 @@ def visit(member, context_description)
 
       def visit_fields(type_defn)
         type_defn.all_fields.each do |field_defn|
-          instrumented_field_defn = @field_instrumenters.reduce(field_defn) do |defn, inst|
-            inst.instrument(type_defn, defn)
-          end
+          visit_field_on_type(type_defn, field_defn)
+        end
+      end
+
+      def visit_field_on_type(type_defn, field_defn, dynamic_field: false)
+        instrumented_field_defn = @field_instrumenters.reduce(field_defn) do |defn, inst|
+          inst.instrument(type_defn, defn)
+        end
+        if !dynamic_field
           @instrumented_field_map[type_defn.name][instrumented_field_defn.name] = instrumented_field_defn
-          @type_reference_map[instrumented_field_defn.type.unwrap.name] << instrumented_field_defn
-          visit(instrumented_field_defn.type, "Field #{type_defn.name}.#{instrumented_field_defn.name}'s return type")
-          instrumented_field_defn.arguments.each do |name, arg|
-            @type_reference_map[arg.type.unwrap.to_s] << arg
-            visit(arg.type, "Argument #{name} on #{type_defn.name}.#{instrumented_field_defn.name}")
-          end
         end
+        @type_reference_map[instrumented_field_defn.type.unwrap.name] << instrumented_field_defn
+        visit(instrumented_field_defn.type, "Field #{type_defn.name}.#{instrumented_field_defn.name}'s return type")
+
+        instrumented_field_defn.arguments.each do |name, arg|
+          @type_reference_map[arg.type.unwrap.to_s] << arg
+          visit(arg.type, "Argument #{name} on #{type_defn.name}.#{instrumented_field_defn.name}")
+        end
+
+        # Construct arguments class here, which is later used to generate GraphQL::Query::Arguments
+        # to be passed to a resolver proc
+        instrumented_field_defn.arguments_class = GraphQL::Query::Arguments.construct_arguments_class(
+          argument_definitions: instrumented_field_defn.arguments,
+        )
```

Comment:
This bit is of code is used for Fields, InputObjects and Directives, do you think it's possible to refactor this into one method, maybe something like 

```ruby 
# This method: 
# - reads `.arguments 
# - constructs a class 
# - assigns it to `.arguments_class=`
GraphQL::Query::Arguments.construct_arguments_class(argument_owner)
```